packages/react-storage/src/components/StorageBrowser/views/LocationActionView/DestinationPicker.tsx

```diff
@@ -0,0 +1,135 @@
+import {
+  getDestinationListFullPrefix,
+  getDestinationPickerTableData,
+} from './utils/getDestinationPickerDataTable';
+import { ControlsContext } from '../../controls/types';
+import { Breadcrumb } from '../../components/BreadcrumbNavigation';
+import { DescriptionList } from '../../components/DescriptionList';
+const {
+  actionSetDestination,
+  actionDestinationPickerCurrentFolderSelected,
+  actionDestinationPickerNoMoreFolders,
+  actionDestinationPickerDefaultError,
+} = displayText;
+
+const DEFAULT_PAGE_SIZE = 10;
+export const DEFAULT_LIST_OPTIONS = {
+  pageSize: DEFAULT_PAGE_SIZE,
+  delimiter: '/',
+};
+
+export const DestinationPicker = ({
+  destinationList,
+  onSetDestinationList,
+}: {
+  destinationList: string[];
+  onSetDestinationList: (destination: string[]) => void;
+}): React.JSX.Element => {
+  const {
+    bucket,
+    items,
+    hasNextToken,
+    currentPage,
+    isLoading,
+    hasError,
+    handleNext,
+    handlePrevious,
+    range,
+  } = useDestinationPicker({ destinationList });
+
+  const handleNavigateFolder = (key: string) => {
+    const newPath = [...destinationList, key.replace('/', '')];
+    onSetDestinationList(newPath);
+  };
+
+  const handleNavigatePath = (index: number) => {
+    const newPath = destinationList.slice(0, index + 1);
+    onSetDestinationList(newPath);
+  };
+
+  const pageItems = React.useMemo(() => {
+    const [start, end] = range;
+    return items.slice(start, end);
+  }, [range, items]);
+
+  const disableNext =
+    !hasNextToken && currentPage * DEFAULT_PAGE_SIZE > items.length;
+  const disablePrevious = currentPage === 1;
+
+  const tableData = getDestinationPickerTableData({
+    items: pageItems,
+    handleNavigateFolder,
+  });
+
+  const contextValue: ControlsContext = {
+    data: {
+      tableData,
+    },
+  };
+
+  const noSubfolders = !items.length;
+  const messageVariant = noSubfolders ? 'info' : 'error';
+  const message = noSubfolders
+    ? `${actionDestinationPickerCurrentFolderSelected} ${getDestinationListFullPrefix(
+        destinationList
+      )}. ${actionDestinationPickerNoMoreFolders}`
+    : actionDestinationPickerDefaultError;
+  const showMessage = !isLoading && (noSubfolders || hasError);
+
+  return (
+    <ControlsContextProvider {...contextValue}>
+      <ViewElement className={`${CLASS_BASE}__action-destination`}>
+        <DescriptionList
+          descriptions={[
+            {
+              term: `${actionSetDestination}:`,
+              details: destinationList.length ? (
+                <>
+                  {destinationList.map((key, index) => (
+                    <Breadcrumb
+                      isCurrent={index === destinationList.length - 1}
+                      key={`${key}-${index}`}
+                      onNavigate={() => handleNavigatePath(index)}
+                      // If bucket level access, show bucket name as root breadcrumb
+                      name={key === '' ? bucket : key.replace('/', '')}
```

Comment:
not a blocker but i think this logic for creating breadcrumb props can go in `useDestinationPicker`. same for Line 85.

Reply:
🍕

Reply:
⭐ suck

Reply:
+1, that we are passing `bucket` through `useDestinationPicker` feels fragile